### ISSUE TITLE
[BUG] feature flag opensearch_security.multitenancy.enabled was broken

### ISSUE THREAD

**Issue by Flyingliuhub** [opensearch-project/security-dashboards-plugin#1412]
**What is the bug?**
the feature flag opensearch_security.multitenancy.enabled was broken when use it in the opensearchdashboard.yaml file, the recently change https://github.com/opensearch-project/security-dashboards-plugin/pull/1394/files will override this flag. 

**How can one reproduce the bug?**
1. opensearch_security.multitenancy.enabled: false in the opensearchdasbhoard.yaml
2. start opensearch dasbhoards
3. it pop up the windows ask to choose tenant
4. cannot choose any tenant, please see the screenshot below

**What is the expected behavior?**
opensearch_security.multitenancy.enabled: false should functional 

**Do you have any screenshots?**
<img width="819" alt="image" src="https://user-images.githubusercontent.com/33105471/235832066-eaed96e4-79da-4230-8446-48306138996c.png">

**Do you have any additional context?**
N/A

**Comment by cwperks** [comment]
One solution could be to use the value in `opensearch_dashboards.yml` as the default value.

For this [line](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/public/apps/account/account-nav-button.tsx#L52) 

```
const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(true);
```

could be changed to:

```
const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(props.config.multitenancy.enabled || false);
```

@Flyingliuhub @kavilla Are you able to dynamically turn off the multitenancy enabled value and have dashboards run as expected?

**Comment by cwperks** [comment]
@abhivka7 What are your thoughts on this issue?

**Comment by cliu123** [comment]
@cwperks The default value should be `true`, like this as Multi-tenancy is enabled by default:
``````
const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(props.config.multitenancy.enabled || true);
```

**Comment by cliu123** [comment]
@cwperks Good point! But the default value should be `true`, like the following as Multi-tenancy is enabled by default:
``````
const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(props.config.multitenancy.enabled || true);

**Comment by cwperks** [comment]
@cliu123 Thank you for chiming in! 

`props.config.multitenancy.enabled || true` would always return true, so this just needs to be updated to use `props.config.multitenancy.enabled` if its not undefined otherwise use the default of `true`.

**Comment by kavilla** [comment]
That would help as it would appear the static config currently does not seem to have much purpose within OpenSearch Dashboards as it would prioritize the OpenSearch dynamic setting over the OpenSearch Dashboards.

However, we will still have the issue with the state being accessed before it is updated from the async function. In this sense we should ensure that we wait for the response from dynamic config prior to accessing and executing logic.

**Comment by DarshitChanpura** [comment]
> @cwperks Good point! But the default value should be `true`, like the following as Multi-tenancy is enabled by default:
> 
> ```
> const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(props.config.multitenancy.enabled || true);
> ```

@cliu123 The default value should be derived from the props where it is set. Providing the default in the file shouldn't be the decider. If we set true in the local state then we are not relying on the config passed in the opensearch_dashboards.yml and thus the value will always remain true, until the call to fetch multi-tenancy data (from backend) is made. This is when the next evaluation happens to check whether the popup should be displayed.

IMO, this is how the default state should be set, unless mutli-tenancy flag will not be defined in opensearch_dashboards.yml at the cluster startup:
```
const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(props.config.multitenancy.enabled || false);
```


@Flyingliuhub @kavilla For customers facing this issue immediately, they should execute the following curl request to update the multi-tenancy config in security index:
```
curl -XPUT https://<usercreds>@URL/_plugins/_security/api/tenancy/config --data '{ "multitenancy_enabled" : false}' -H "Content-type: application/json" | jq
```
_NOTE: Use admin creds and `--insecure` option if running on localhost_

I have created a PR to fix this issue #1417

**Comment by abhivka7** [comment]
We wanted to remove complete dependancy from` opensearch_security.multitenancy.enabled in opensearchdashboard.yaml `file.

But I missed to remove that dependancy here: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/plugin.ts#L132.
When `opensearch_security.multitenancy.enabled` is disabled then multitenant routes are not setup and user was getting 404 error. 

Going forward, we want to go with one single source of truth approach. Therefore I am raising a PR soon to remove dependancy from ` opensearch_security.multitenancy.enabled in opensearchdashboard.yaml `file.

**Comment by abhivka7** [comment]
I have raised a PR for fixing this issue: https://github.com/opensearch-project/security-dashboards-plugin/pull/1419
Please review.

**Comment by kavilla** [comment]
There appears to be other issues with the current implementation.

The [docs](https://opensearch.org/docs/latest/security/multi-tenancy/multi-tenancy-config/) say: 
```
opensearch_security.multitenancy.enabled 	Enables or disables multi-tenancy in OpenSearch Dashboards. Default is true.
```

But looking at the config the default value is `false`. When we release a build we do set the config but anyone who pulls the source or min artifact can simply not set that value and the OpenSearch Dashboards side will be defaulted to `false`. While OpenSearch appears that [default value is set to true](https://github.com/search?q=repo%3Aopensearch-project%2Fsecurity%20multitenancy_enabled&type=cod). So the code in the `public` folder will utilize the dynamic config to determine to set up multitenancy but the code in the `server` folder will use the static config to determine if it should set up things related to multitenancy, to which some parts of the public folder expect to be set up.

For example, the [OSD plugin will not setup the index template](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/plugin.ts#L160) if the static OSD config is set to false.

Which in this context isn’t the biggest deal because if the static config is set to false the user will not be able to dynamically configure their multitenancy. This is due to the [routes depend on the static config](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/plugin.ts#L132) being `true` to be setup.

But it still makes for a bad experience because I just get a bunch of errors because the `public` client side will render the logic as if multitenancy is enabled while the server side will not execute the logic since the static config said multitenancy is disabled. You can see this by commenting on the config if using the release, starting the application, and navigate to the `Security > Tenants > Configure` and try to make any changes and save here. You will get errors.

In this feature's current state, we will have to be explicit that users can not use the dyanmic config if they have disabled the static config in OSD.

However we can not simply just make the code changes so that the `server` side respect the dynamic config because the routes and index template are currently only set up at security dashboards plugin discovery which only occurs at OSD startup. If we were to do that without refactor other parts, an admin would be able to update the configuration dynamically and then bounced OSD which will then trigger the re-discovery and not setup the index templates. One proposal would be if the multitenancy is dynamic then we should ensure that we always set up these routes regardless and just reject requests if multitenancy is disabled. We can also consider moving the index template creation outside of the startup process.

**Comment by abhivka7** [comment]
@kavilla With https://github.com/opensearch-project/security-dashboards-plugin/pull/1419, we will tie up every tenancy feature in OpenSearch Dashboards with feature flag `opensearch_security.multitenancy.enabled of opensearchdashboard.yaml` file. Therefore the routing problem like the one in `Security > Tenants > Configure` will go away because we won't show users the configure tab if their `opensearch_security.multitenancy.enabled ` is disabled.

Similarly, all other tenancy related popups (like choose tenants during first login) and features ( like switch tenant option in account-nav-bar and `Security > Tenants ' page ) will be disabled if we disable `opensearch_security.multitenancy.enabled `.

**Comment by abhivka7** [comment]
The PR for fixing this issue is merged. We can close this issue.

**Comment by norsco15** [comment]
Hello,

I'm facing a similar issue to [#1412], where I'm unable to use the multitenancy feature due to an authorization exception and "no permissions" error.

The feature flag opensearch_security.multitenancy.enabled is set to true in the `opensearch_dasbhoards.yml` file, but it's not working as expected.

To reproduce the issue, I did the following:

- Set opensearch_security.multitenancy.enabled: true in the `opensearch_dasbhoards.yml` file.
- Started opensearch dashboards.
- I am prompted with a window asking to choose a tenant.
- However, I cannot choose any tenant (similar to the screenshot in the original issue), also I cannot visualize anything due to the following error ` log   [13:18:10.986] [error][data][opensearch] [security_exception]: no permissions for [indices:data/read/search] and User [name=client1, backend_roles=[client1_role], requestedTenant=null]`.

The expected behavior is for the opensearch_security.multitenancy.enabled: true to allow multitenancy functionality.

The issue seems to persist even after the fix [#1419] was merged.

Please advise on any possible solutions or workarounds for this issue.

Thank you.

**Comment by DarshitChanpura** [comment]
@norsco15 Could you post the output of API call: `_plugins/_security/api/tenancy/config`? And if possible could you also post your user and roles config (barring any sensitive info) as this would be helpful in debugging.
